### PR TITLE
Fix ability to purge value of a custom field

### PR DESCRIPTION
**Pull request by vboctor**
Fixes #20002

### CODE REVIEW

File: core/custom_field_api.php [hunk 1]
```diff
@@ -1269,12 +1269,31 @@ function print_custom_field_input( array $p_field_def, $p_bug_id = null ) {
...
 		call_user_func( $g_custom_field_type_definition[$p_field_def['type']]['#function_print_input'], $p_field_def, $t_custom_field_value );
+		echo '<input type="hidden" name="' . custom_field_presence_field_name( $p_field_def['id'] ) . '" value="1" />';
 	} else {
 		trigger_error( ERROR_CUSTOM_FIELD_INVALID_DEFINITION, ERROR );
 	}
 }
 
 /**
+ * Constructs the name of the field that used as a flag that a custom field is present on the form.`
+ * @param $p_custom_field_id  The custom field id to create the field name for.
```
Comment (atrol) [inline core/custom_field_api.php]: @param integer $p_custom_field_id

File: core/custom_field_api.php [hunk 2]
```diff
@@ -1269,12 +1269,31 @@ function print_custom_field_input( array $p_field_def, $p_bug_id = null ) {
...
+ * @return string The field name.
+ */
+function custom_field_presence_field_name( $p_custom_field_id ) {
+	return 'custom_field_' . (int)$p_custom_field_id . '_presence';
+}
+
+/**
+ * Checks the presence of a custom field on the form.
+ * @param $p_custom_field_id  The custom field id to check.
```
Comment (atrol) [inline core/custom_field_api.php]: @param integer $p_custom_field_id

File: core/custom_field_api.php [hunk 3]
```diff
@@ -1269,12 +1269,31 @@ function print_custom_field_input( array $p_field_def, $p_bug_id = null ) {
 	global $g_custom_field_type_definition;
 	if( isset( $g_custom_field_type_definition[$p_field_def['type']]['#function_print_input'] ) ) {
 		call_user_func( $g_custom_field_type_definition[$p_field_def['type']]['#function_print_input'], $p_field_def, $t_custom_field_value );
+		echo '<input type="hidden" name="' . custom_field_presence_field_name( $p_field_def['id'] ) . '" value="1" />';
```
Comment (atrol) [inline core/custom_field_api.php]: Use function print_hidden_input?

**Comment by vboctor** [comment]
@atrol I've applied your feedback.

**Comment by atrol** [comment]
+1 in principle, no time to test

File: core/custom_field_api.php [hunk 4]
```diff
@@ -1269,12 +1269,31 @@ function print_custom_field_input( array $p_field_def, $p_bug_id = null ) {
...
 	if( isset( $g_custom_field_type_definition[$p_field_def['type']]['#function_print_input'] ) ) {
 		call_user_func( $g_custom_field_type_definition[$p_field_def['type']]['#function_print_input'], $p_field_def, $t_custom_field_value );
+		print_hidden_input( custom_field_presence_field_name( $p_field_def['id'] ), '1' );
 	} else {
 		trigger_error( ERROR_CUSTOM_FIELD_INVALID_DEFINITION, ERROR );
 	}
 }
 
 /**
+ * Constructs the name of the field that used as a flag that a custom field is present on the form.`
```
Comment (dregad) [inline core/custom_field_api.php]: Suggested rewording:
Constructs the name of a field used as a flag to indicate that a custom field is present on the form

File: core/custom_field_api.php [hunk 5]
```diff
@@ -1269,12 +1269,31 @@ function print_custom_field_input( array $p_field_def, $p_bug_id = null ) {
...
+function custom_field_presence_field_name( $p_custom_field_id ) {
+	return 'custom_field_' . (int)$p_custom_field_id . '_presence';
+}
+
+/**
+ * Checks the presence of a custom field on the form.
+ * @param integer $p_custom_field_id  The custom field id to check.
+ * @return bool true when field is on form, false otherwise.
+ */
+function custom_field_present( $p_custom_field_id ) {
```
Comment (dregad) [inline core/custom_field_api.php]: Suggest renaming function to `custom_field_is_present`

**Comment by dregad** [comment]
See my [comment in the tracker](https://mantisbt.org/bugs/view.php?id=20002#c51319)

**Comment by vboctor** [comment]
@dregad I've addressed your comment.  Merged.

**Comment by dregad** [comment]
Does that cover my remark in #20002?

**Comment by vboctor** [comment]
@dregad yes, the scenario about by-passing access level when custom field is not on the form was removed.

**Comment by dregad** [comment]
Cheers

**Pull request merged**